diff -r d1f8610cdffd configs/example/fs.py
--- a/configs/example/fs.py	Fri Jan 15 11:30:13 2016 +0000
+++ b/configs/example/fs.py	Sun Jan 17 19:11:50 2016 +0100
@@ -123,6 +123,9 @@
                                              voltage_domain =
                                              test_sys.cpu_voltage_domain)
 
+    if options.disk_image is not None:
+        test_sys.disk0.childImage(disk(options.disk_image))
+
     if options.kernel is not None:
         test_sys.kernel = binary(options.kernel)
 
