diff -r a22c4bac7e15 configs/common/FSConfig.py
--- a/configs/common/FSConfig.py	Sat May 07 14:43:06 2016 -0400
+++ b/configs/common/FSConfig.py	Sun Jun 12 22:35:43 2016 +0200
@@ -169,7 +169,7 @@
     self.partition_desc.port = self.membus.master
     self.intrctrl = IntrControl()
     self.disk0 = CowMmDisk()
-    self.disk0.childImage(disk('disk.s10hw2'))
+    self.disk0.childImage(mdesc.disk())
     self.disk0.pio = self.iobus.master
 
     # The puart0 and hvuart are placed on the IO bus, so create ranges
